[primTorch] Add new test modes for refs and aliases

nkaretnikov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #84957



cc @ezyang @mruberry @ngimel @Lezcano @peterbell10

nkaretnikov committed:
[primTorch] Test ref aliases, disallow aliases in `python_ref_db`

[ghstack-poisoned]

nkaretnikov commented:
```
[gw0] FAILED test/onnx/test_pytorch_onnx_no_runtime.py 
 test/onnx/test_pytorch_onnx_no_runtime.py::TestONNXExport.test_source_range_propagation ⨯98% █████████▉
 
   File "/opt/conda/lib/python3.7/site-packages/torch/onnx/symbolic_opset9.py", line 2491, in layer_norm
    g, input, normalized_shape, weight, bias, eps, cudnn_enable, False
  File "<@beartype(torch.onnx.symbolic_opset9._layer_norm_returns_normalized_input_mean_rstd) at 0x7f5eb95e9f80>", line 182, in _layer_norm_returns_normalized_input_mean_rstd
  File "/var/lib/jenkins/.local/lib/python3.7/site-packages/beartype/_decor/_error/errormain.py", line 302, in raise_pep_call_exception
    f'{exception_prefix}violates type hint {repr(hint)}, as '
beartype.roar.BeartypeCallHintReturnViolation: @beartyped _layer_norm_returns_normalized_input_mean_rstd() return "6 defined in (%6 : Tensor = aten::ATen[cudnn_enable=1, eps=1.0000000000000001e-05, normalize... violates type hint typing.Tuple[torch.Value, typing.Union[torch.Value, NoneType], typing.Union[torch.Value, NoneType]], as "6 defined in (%6 : Tensor = aten::ATen[cudnn_enable=1, eps=1.0000000000000001e-05, normalize... not instance of tuple.
```

might be flaky, passes locally for me

nkaretnikov committed:
Update on "[primTorch] Test ref aliases, disallow aliases in `python_ref_db`"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Test ref aliases, disallow aliases in `python_ref_db`"




[ghstack-poisoned]

lezcano reviewed (approved):
LGTM

Lezcano commented on test/test_ops.py:
```diff
@@ -1672,7 +1697,16 @@ class TestRefsOpsInfo(TestCase):
+        # skip aliases since we don't want them defined in the ref db
+        if op in self.ref_alias_names_set:
+            return
+        self.assertIn(op, self.ref_db_names_set)
```
It may be good to add a message here explaining what to do if this test fails.

mruberry commented on test/test_ops.py:
```diff
@@ -227,17 +227,26 @@ def _ref_test_helper(self, ctx, device, dtype, op, skip_zero_numel=False, skip_z
+                            self.assertEqual(a._is_view(), b._is_view())
 
-            # Computes the dtype the more precise computatino would occur in
+            # Computes the dtype the more precise computation would occur in
```
Nice spelling correction

mruberry commented on test/test_ops.py:
```diff
@@ -227,17 +227,26 @@ def _ref_test_helper(self, ctx, device, dtype, op, skip_zero_numel=False, skip_z
             ):
                 continue
+
+            # Note: this deliberately uses 'op.torch_opinfo' in both cases
```
I think this comment could be a little clearer -- what's going on in this block, and what "cases" are being described?

mruberry commented on test/test_ops.py:
```diff
@@ -227,17 +227,26 @@ def _ref_test_helper(self, ctx, device, dtype, op, skip_zero_numel=False, skip_z
             with ctx():
-                ref_result = op(sample.input, *sample.args, **sample.kwargs)
-            torch_result = op.torch_opinfo(sample.input, *sample.args, **sample.kwargs)
+                ref_results = [f(sample.input, *sample.args, **sample.kwargs)
```
I think it'd be clearer to enumerate in a loop outside and do each comparison directly vs. gathering the comparisons. It would also make it easier to provide an error message in the future, because the error would be thrown within the loop for a particular alias

mruberry commented on test/test_ops.py:
```diff
@@ -227,17 +227,26 @@ def _ref_test_helper(self, ctx, device, dtype, op, skip_zero_numel=False, skip_z
+                ref_results = [f(sample.input, *sample.args, **sample.kwargs)
+                               for f in itertools.chain((op,), aliases)]
+            torch_results = [f(sample.input, *sample.args, **sample.kwargs)
+                             for f in itertools.chain((op.torch_opinfo,), aliases)]
```
How does this work? Isn't the `chaining` done with the same `aliases` in both cases, meaning the the same thing is being compared?

nkaretnikov replied:
thanks for bringing this up, there are multiple issues here. apologies for the long comment, i've tried to explain it as brief as i can.

## about the context

for instance, with `TorchRefsMode(strict=True)` as `ctx`, a ref with `torch.foo` in it will call `_refs.foo` instead (this is already documented in `test_python_ref`). more details below.

## testing the op against the ref

let's first look at how we test the op against its ref. when we test `op` against `op.torch_opinfo` here, we call `_refs.foo` (that's `op`, but it's in the context, so all inner `torch` calls in the ref function body are mapped to ref impls as described above) against `torch.foo` (that's `op.torch_opinfo`, but it's outside of the context, so execution continues as in regular pytorch).

this is a good test because we check different entrypoints, but i now realize it's also not great because users will never call `_refs` functions directly. iiuc, they will just write regular pytorch code inside a new context, maybe like this:

```py
with primtorch():  # made-up future context for primtorch
  torch.log_softmax(...)
  # the rest of the code that should be copy-pastable from regular pytorch
```

_i think we need to add another pair of input functions where we check `op.torch_opinfo` against itself inside and outside the context_ because that's what users will actually be executing (if my understanding is correct).

## testing aliases

with aliases, we just call stuff in `torch` in both `ref_results` and `torch_results`. but in `ref_results`, inner calls to `torch` are translated to `_refs` calls instead (due to it being in the context).

for instance, we might have `torch.nn.functional.log_softmax` as an alias in both cases:

```py
def log_softmax(input: Tensor, dim: Optional[int] = None, _stacklevel: int = 3, dtype: Optional[DType] = None) -> Tensor:
    if has_torch_function_unary(input):
        # [1]
        return handle_torch_function(log_softmax, (input,), input, dim=dim, _stacklevel=_stacklevel, dtype=dtype)
[...]
    return ret
```

in the context, the `log_softmax` call at `[1]` will go to `_refs.nn.functional.log_softmax`.

_perhaps we could try adding another set of inputs where we'd test `_refs.alias` against `torch.alias`, but we have no convenient way of getting to `_refs.alias` as of now. and based on the assumption above, those will never be executed directly by users, so it's seems not worth it._

## summary

rather than doing anything about the aliases, i'm thinking about adding a test where i test `op.torch_opinfo` against itself (as described above) in addition to the existing tests.

thoughts?

nkaretnikov replied:
i also forgot to mention that most aliases should be defined as `foo = torch.foo` anyway, so it shouldn't matter much in practice(?). there might be some corner cases, like where we have slightly more complicated wrappers for `nn.functional` and `special` variants due to signature incompatibilities, but i don't feel like this justifies complicating this test. and again, the entrypoints that we need to be testing are `torch.foo` (inside and outside the context), iiuc

mruberry replied:
Yes we need to make sure the testing is correctly remapping ops, and that's the most important thing. Maybe we should have a separate test that just checks that the remapping is correct? 

For aliases, there are "true" aliases, which we should define as `foo = torch.foo`, and "near" aliases, as you're pointing out. I'm not that worried about testing either except that it'd be nice to verify both call the aliased operation properly. If that's too tricky to figure out then let's not worry about it for now, it should be easy to implement later.

nkaretnikov replied:
rewrote the comment and introduced new test modes

mruberry commented on test/test_ops.py:
```diff
@@ -316,7 +327,7 @@ def _distance(a, b):
     # Tests that experimental Python References perform the same computation
     # as the operators they reference, when operator calls in the torch
-    # namesapce are remapped to the refs namespace (torch.foo becomes refs.foo).
+    # namespace are remapped to the refs namespace (torch.foo becomes refs.foo).
```
Nice spelling correction

mruberry commented on test/test_ops.py:
```diff
@@ -1561,7 +1572,20 @@ class TestRefsOpsInfo(TestCase):
     ref_ops_names = tuple(itertools.chain.from_iterable(
         [f"{path}.{op}" for op in module_all] for path, module_all in module_alls))
-    ref_db_names = set(ref_op.name for ref_op in python_ref_db)
+    # 'set' filters out repeated ref names (such as those with variants)
```
Nice comment

mruberry commented on test/test_ops.py:
```diff
@@ -1561,7 +1572,20 @@ class TestRefsOpsInfo(TestCase):
+    ref_db_names_set = set(ref_op.name for ref_op in python_ref_db)
+    # sort to always use the same order
+    ref_db_names = sorted(ref_db_names_set)
+    # Note: this deliberately uses 'op.torch_opinfo' because 'op.aliases' is
```
Should `op.aliases` be empty, though, as we think about how to test aliases?

mruberry commented on test/test_ops.py:
```diff
@@ -1561,7 +1572,20 @@ class TestRefsOpsInfo(TestCase):
+    # prefix.  Those are skipped here since it's not clear how to check
+    # those yet.
+    ref_alias_names_set = set(
+        "_refs." + alias.name
```
It seems weird that this is doing string concatenation here

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16160,11 +16174,13 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
     ElementwiseUnaryPythonRefInfo(
         "_refs.lgamma",
         torch_opinfo_name="lgamma",
+        test_aliases=False,  # RuntimeError: no _refs support for torch.special.gammaln
```
Instead of introducing new metadata, what about xfailing the alias test?

nkaretnikov replied:
don't want to xfail the entire test because it would be a regression in testing coverage (all current ref/torch checks should pass as is). also don't want to split out alias testing into a separate test (which we could xfail) because it would require some refactoring and i don't want to change this too much as it's originally written by nvfuser people (iiuc)

mruberry replied:
Testing aliases separately sounds like the right approach. Our nvfuser colleagues can help review if necessary

mruberry reviewed (commented):
Hey @nkaretnikov! I made some inline comments for your review. Basically I wonder if we should put the alias info into the reference's opinfo entry? Maybe it can be autopopulated by from the original reference? There a couple other ideas for your review, too.

Looking forward to hearing your thoughts!

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov commented:
@lezcano @mruberry @ngimel Addressed feedback, CI is green, PTAL at this whole stack.

Summary of this PR:
- I moved checks from [the third PR](https://github.com/pytorch/pytorch/pull/84792) here.
- I introduced ref aliases, which point directly to `_refs.foo` implementations. This allows to call these directly as well as to test them in various contexts.
- I introduced new context test modes (see `CtxTestMode`). The _most important_ ones are `OP_OP` and `OP_ALIASES`. Any new test failures with these two tests (not present in other modes) highlight coverage or API compat problems because it means that  a `torch.foo` call in `TorchRefsMode(strict=True)` failed. IIUC, this corresponds to how users will actually use PrimTorch (also see [this comment](https://github.com/pytorch/pytorch/pull/84502#discussion_r963578734)). The other two `REF` modes exist for additional checking. We should pay attention to those as well, but they are less critical. See the comment in ` _ref_test_helper` explaining the differences between these modes.

```py
with primtorch():  # made-up context
  torch.foo()  # NOT torch._refs.foo()
```

Some examples of API compat problems:
```
# TypeError: margin_ranking_loss() got an unexpected keyword argument 'size_average'
DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_mode_op_op'),
            
# TypeError: hinge_embedding_loss() got an unexpected keyword argument 'size_average'
DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_mode_op_op'),
```

An example of a coverage issue:
```
# RuntimeError: no _refs support for torch.Tensor.unfold
DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_mode_op_op'),
```

nkaretnikov commented:
This is out of scope for this PR, but now that I wrote the above, I should mention that it seems surprising to me that `TorchRefsMode(strict=True)` even executes the original implementation at all. It only replaces calls _inside the function body_, not _the first call_ in the context. Are people aware of this?

```py
# orig impl
def elu(input: Tensor, alpha: float = 1.0, inplace: bool = False) -> Tensor:
    print("ORIG ELU")
    if has_torch_function_unary(input):
        print("BEFORE ORIG ELU CALL")
        return handle_torch_function(elu, (input,), input, alpha=alpha, inplace=inplace)
[...]

# ref
[...]
def elu(
    a: TensorLikeType, alpha: Optional[NumberType] = None, inplace: bool = False
) -> TensorLikeType:
    print("REF ELU")
```

```py

In [2]: import torch

In [3]: from torch._prims.context import TorchRefsMode

In [4]: with TorchRefsMode(strict=True):
   ...:     torch.nn.functional.elu(input=torch.Tensor([7.]))
   ...: 
ORIG ELU
BEFORE ORIG ELU CALL
REF ELU
```

mruberry commented:
> This is out of scope for this PR, but now that I wrote the above, I should mention that it seems surprising to me that `TorchRefsMode(strict=True)` even executes the original implementation at all. It only replaces calls _inside the function body_, not _the first call_ in the context. Are people aware of this?
> 
> ```python
> # orig impl
> def elu(input: Tensor, alpha: float = 1.0, inplace: bool = False) -> Tensor:
>     print("ORIG ELU")
>     if has_torch_function_unary(input):
>         print("BEFORE ORIG ELU CALL")
>         return handle_torch_function(elu, (input,), input, alpha=alpha, inplace=inplace)
> [...]
> 
> # ref
> [...]
> def elu(
>     a: TensorLikeType, alpha: Optional[NumberType] = None, inplace: bool = False
> ) -> TensorLikeType:
>     print("REF ELU")
> ```
> 
> ```python
> In [2]: import torch
> 
> In [3]: from torch._prims.context import TorchRefsMode
> 
> In [4]: with TorchRefsMode(strict=True):
>    ...:     torch.nn.functional.elu(input=torch.Tensor([7.]))
>    ...: 
> ORIG ELU
> BEFORE ORIG ELU CALL
> REF ELU
> ```

That is very interesting and not what I'd expect. cc @ezyang

ezyang commented:
Everything looks in order to me... TorchFunctionMode doesn't work by monkey-patching all of the functions in torch namespace, it works via that conditional you see which detects if mode is enabled and goes inside if so. So yes, we have to go to the function first, hit the conditional, and then hit the ref handler.

mruberry commented on test/test_ops.py:
```diff
@@ -69,6 +70,29 @@
 
+
+@enum.unique
+class CtxTestMode(enum.Enum):
```
Add a comment describing what this enum is does and how it's intended to be used

mruberry commented on test/test_ops.py:
```diff
@@ -69,6 +70,29 @@
+    OP_ALIASES = enum.auto()
+
+
+class CtxTestInput(object):
```
Add a comment for this class, too

mruberry commented on test/test_ops.py:
```diff
@@ -237,89 +262,132 @@ def _ref_test_helper(
-                    raise e
 
-                ex = e
+            # Check op behavior and API compatibility inside and outside ctx.
```
So the current test compares the `torch.foo` vs. `torch._refs.foo`, right? And this extends testing to include:

1.  redirecting from `torch.foo` to `torch._refs.foo` via `torch_function`
2. testing alias consistency

Today in `test_ops.py` we test alias consistency in a very time-consuming way here:

https://github.com/pytorch/pytorch/blob/70cce9f8d1a099af8b017d7263897d3ca2fb9fe6/test/test_ops.py#L867

When really we'd like to test aliases by just asking if `foo_alias is foo`. Of course, as we've discussed briefly in these notes there are also "near aliases" that wouldn't satisfy the above test, but we should probably just treat those as separate operations or fix them so they're actual aliases. Note that for `torch.*` aliases it's not true that `torch.foo_alias is foo` because of how we decided to create alias for operations (they have autogenerated bindings from native_functions.yaml), but Python references don't use yaml or C++, and so could be defined to satisfy the `is` expression, as this PR does.

Anyway, back to these text extensions. The first part of the test that validates the `torch_function` mapping is correct seems more like we're testing that mapping than anything, and I imagine that's already done or best done elsewhere? Further, shouldn't we validate that mapping by ensuring the mapped-to operation is correct, rather than using blackbox testing? 

As for aliases, a separate test for Python Reference Alias Consistency is fine. It could start by asking if `alias_foo is foo` and if that's not true then it could assume that the alias is a "near alias" that should be blackbox tested for input->output consistency. 

How does that sound?

mruberry commented on test/test_ops.py:
```diff
@@ -1646,27 +1742,28 @@ class TestRefsOpsInfo(TestCase):
         '_refs.narrow',
         '_refs.nn.functional.l1_loss',
-        '_refs.nn.functional.log_softmax',
+        '_refs.nn.functional.layer_norm',
```
Nice alphabetizing

mruberry commented on test/test_ops.py:
```diff
@@ -1688,12 +1785,40 @@ class TestRefsOpsInfo(TestCase):
+            return
+        self.assertIn(op, self.ref_ops_names)
+
+    @parametrize("op", ref_db_names)
```
I don't think I'd parametrize this test -- it's just asserting a property of the entire sequence. It can fail and the message can describe which aliases are redundantly OpInfo'd

nkaretnikov replied:
Why not? I feel like with the decorator it's more clear which part of the test is actually important. There's less boilerplate and it doesn't take that much time to run. The other tests in the class also use the decorator, so it's all consistent.

lezcano replied:
Multiply the "it doesn't take that much time to run" by the number of CI jobs times the number of commits time the number of PRs and...
I agree with Mike on this one. A better way of doing this is working with the whole set and report all the errors at once in one test.

mruberry commented on test/test_ops.py:
```diff
@@ -1688,12 +1785,40 @@ class TestRefsOpsInfo(TestCase):
 
     @parametrize("op", ref_ops_names)
     def test_refs_are_in_decomp_table(self, op):
+        # skip aliases since those will be registered via their respective ops
```
Like with the above I don't think this test needs to be parametrized?

nkaretnikov replied:
This one iterates over each op. It would be messier without the decorator, IMO.

mruberry commented on torch/_refs/linalg/__init__.py:
```diff
@@ -56,6 +58,18 @@ def check_norm_dtype(dtype: Optional[torch.dtype], x_dtype: torch.dtype, fn_name
 from torch._decomp import register_decomposition
 
 
+# TODO: Defined as an alias in the OpInfo, but has different default arguments.
```
Agreed these should not be aliases

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16744,9 +16793,20 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
         # https://github.com/pytorch/pytorch/issues/85258
         supports_nvfuser=False,
+        skips=(
+            # RuntimeError: "round_vml_cpu" not implemented for 'Short', 'Int', 'Long', 'Char', 'Byte'
```
Why did these failures start when the alias was added? 

According to the documentation integer inputs should just return a copy of themselves:

https://pytorch.org/docs/master/generated/torch.round.html

nkaretnikov replied:
Filed an issue: https://github.com/pytorch/pytorch/issues/86326

I'm assuming we still want to rely on `test_variant_consistency_eager` to check the aliases and not check them extensively here even though it means having poor coverage for now (due to the above issue), correct?

mruberry reviewed (commented):
The addition of the aliases looks great, @nkaretnikov. I wonder if we can simplify the testing? Our tests take a long time to run as it is. 

Looking forward to hearing your thoughts

nkaretnikov commented on test/test_ops.py:
```diff
@@ -1688,12 +1785,40 @@ class TestRefsOpsInfo(TestCase):
-            raise unittest.SkipTest(f"{op} does not have an entry in python_ref_db")
-        self.assertIn(op, self.ref_db_names)
+            raise unittest.SkipTest(err)
+        # skip aliases since we don't want them defined in the ref db
```
This talks about separate `*RefInfo` entries for aliases. If we add those, there'll be a lot of repetition (e.g., due to test failures). I think it's better to model this as we do in `OpInfo`s: aliases are just listed as part of the corresponding op entry (which is what this PR does for refs by adding `_refs.foo` aliases).

mruberry replied:
Yeah that sounds good, but the comment may want to include some of that thinking

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov commented:
@lezcano @mruberry PTAL.

Note: the dynamo failure is unrelated, likely introduced by "Re-enable torchdynamo tests (#86297)", which is now reverted on master.

Summary:
- The `REF_OP` mode is _the same_ as before: references are called directly and are tested against torch ops.
- _Optimized_: the `REF_OP_ALIASES` mode only tests "near aliases". At the moment, these are: `diagonal`, `log_softmax`, `softmax`. I think it's better to have this test rather than just fixing these cases for now since we might bump into other "near aliases" in the future.
- _Optimized_: the `OP_OP` and `OP_ALIASES` modes are only executed in the "mock" context. This is a new parameter of `TorchRefsMode` which replaces the ref function body with a sentinel value `MOCK_REF_RESULT`. This is used to test the op to ref dispatch; after that the test exists, which is pretty fast. This discovered API compatibility problems in `margin_ranking_loss` and `hinge_embedding_loss`.
- _Removed_ `maximum` and `minimum` aliases from `max` and `min` `OpInfo`s. These are "subsets" of their respective ops, not "proper aliases".
- Misc: _Removed_ the `parametrize` decorator from tests that could be written without it. _Added_ an additional check against `op_alias_names` in `test_ref_db_has_no_aliases` since the previous version only considered ref aliases. _Added_ documentation to classes.

nkaretnikov commented:
@lezcano @mruberry 2 workdays since my last ping, so a gentle reminder to take a look at this one.

mruberry commented:
@lezcano's going to review first, from our brief offline sync

lezcano commented:
Yes, sorry, I've been incredibly busy this week. I'll go over this one together with Nikita early next week.

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

lezcano commented on test/test_ops.py:
```diff
@@ -485,6 +390,84 @@ def _distance(a, b):
+    # corresponding op (outside of the context). This test calls *different*
+    # entrypoints. Users are not expected to call the ref directly, this is only
+    # done to check for compatibility issues.
+    def _make_test_inputs_ref_op(self, op):  # default
```
these ops can be defined inside the tests themselves. This is better because it limits their scope, puts them close to where people may want to find them, and also would not need to take a `self` explicitly, but could just capture them.

lezcano commented on test/test_ops.py:
```diff
@@ -485,6 +390,84 @@ def _distance(a, b):
+    # example, with TorchRefsMode(strict=True) as the context, a function with
+    # torch.foo in it will call refs.foo instead.
+
+    # Setters
+    # Test different inputs inside and outside an execution context.
+    def _set_test_input_different(self, ctx, no_ctx):
+        # keep in sync with get functions
+        return (ctx, no_ctx)
+
+    # Test the same input inside and outside an execution context.
+    def _set_test_input_same(self, x):
+        # keep in sync with get functions
+        return (x, x)
+
+    # Getters
+    def _get_test_input_ctx(self, test_input):
+        # keep in sync with set functions
+        # (ctx, no_ctx)
+        return test_input[0]
+
+    def _get_test_input_no_ctx(self, test_input):
+        # keep in sync with set functions
+        # (ctx, no_ctx)
+        return test_input[1]
```
There must be a less verbose way of doing this.

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov commented:
@mruberry Mario's feedback addressed, PTAL

lezcano commented:
I'll put together a peoper review sometime next week. I haven't had the time this week.

lezcano commented on test/test_ops.py:
```diff
@@ -279,106 +280,198 @@ def _ref_test_helper(
+                        # sentinel value, so move to the next sample instead of
+                        # doing any consistency checks against the no_ctx
+                        # result.
+                        self.assertEqual(ref_result, MOCK_REF_RESULT)
```
Rather than the comment above, could you add a short message that the dev would get if this assert fails summarising what's the error and what should they do?

nkaretnikov replied:
Done. I also noticed this assert is not always reached because dispatch may fail earlier, so I added exception handling there, which prints a user-friendly error message on failure.

lezcano commented on test/test_ops.py:
```diff
@@ -279,106 +280,198 @@ def _ref_test_helper(
+
+                # Computes the dtype the more precise computation would occur in
+                precise_dtype = torch.bool
+                if prims.utils.is_integer_dtype(dtype):
+                    # Note: bool and integer dtypes do not have more
+                    # precise dtypes -- they simply must be close
+                    precise_dtype = dtype
```
Not for this PR, but we may want to perform the computation on int64 in the int-but-not-bool case.

lezcano commented on test/test_ops.py:
```diff
@@ -279,106 +280,198 @@ def _ref_test_helper(
             msg = "Test passed because the reference was more accurate than the torch operator."
             warnings.warn(msg)
 
+    # Directly call each ref (in the context) and test it against the
+    # corresponding op (outside of the context). This test calls *different*
+    # entrypoints. Users are not expected to call the ref directly, this is only
+    # done to check for compatibility issues.
+    def _make_test_inputs_ref_op(self, op):  # default
+        # (ctx, no_ctx)
+        return ((op, op.torch_opinfo),)  # ref, op
+
```
If this is the standard way to call things (as it's called within a different test), it may be worth it to make it the default value for `make_test_inputs`.

lezcano commented on test/test_ops.py:
```diff
@@ -279,106 +280,198 @@ def _ref_test_helper(
+        # directly, this is only done to check for compatibility issues. This
+        # test *performs* numerical checking.
+        def _make_test_inputs_ref_op_aliases(op):
+            # The output of zip is determined by the shortest input iterable, so
+            # make sure both inputs have the same length.
+            self.assertTrue(len(op.aliases) == len(op.torch_opinfo.aliases),
+                            msg="Ref and op aliases have different lengths")
```
Good one. The default behaviour of `zip` is just terrible.

lezcano commented on test/test_ops.py:
```diff
@@ -279,106 +280,198 @@ def _ref_test_helper(
+        self._ref_test_helper(lambda: TorchRefsMode(strict=True), device, dtype, op,
+                              make_test_inputs=_make_test_inputs_ref_op_aliases)
+
+    # Only checks API compatibility, so test on one device with any dtype.
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
```
Why do we need to skip this test under ASAN if it's not executing any op?

lezcano commented on test/test_ops.py:
```diff
@@ -1763,22 +1861,84 @@ class TestRefsOpsInfo(TestCase):
+
+    # like 'test_refs_are_in_python_ref_db' but check in the other direction,
+    # this includes aliases since those are not allowed in 'ref_db_names'
+    @parametrize("op", itertools.chain(ref_db_names, ref_alias_names))
+    def test_python_refs_are_in_ops_names(self, op):
+        # TODO: skip nvfuser refs since those don't have __all__ and it's not
+        # clear at the moment what would be the best way to check those
+        # https://github.com/pytorch/pytorch/issues/85081
+        if op.startswith("ops.nvprims"):
+            raise unittest.SkipTest(f"Skipping nvfuser ref: {op}")
+        self.assertIn(op, self.ref_ops_names)
```
I think this test may be better not to be parametrised, and simply have it execute once and return a list of ops that are problematic, together with a nice message explaining why did it fail and how to fix it, similar to the test following it.

lezcano commented on test/test_ops.py:
```diff
@@ -1763,22 +1861,84 @@ class TestRefsOpsInfo(TestCase):
+                if op_impl not in torch._decomp.decomposition_table.values():
+                    missing_decomps.append(ref)
+
+        self.assertTrue(extra_decomp_skips == [],
```
Better practice: Either `len(extra_decomps) == 0` or `not extra_decomps`.

lezcano reviewed (commented):
I left a few minor points, but overall looks good. @mruberry could you have a look at this one as well?

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1348,13 +1349,22 @@ def disable_functorch():
+        with _pop_mode_temporarily():
+            yield
+    else:
+        yield
```
`_no_torch_function_mode` was removed in https://github.com/pytorch/pytorch/pull/86458, so I define my own helper. In that PR, see `handle_torch_function` for how the new API is used. `_pop_mode_temporarily` has no check for whether the stack is empty, only an assert on the CPP side, so I check if the mode is enabled to avoid popping an empty stack.

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add new test modes for refs and aliases"




[ghstack-poisoned]

nkaretnikov commented:
@mruberry PTAL
